### PR TITLE
Add CSRF protection for the AJAX section update actions in pegasus

### PR DESCRIPTION
**Pull request by philbogle**
This change adds cross-site request forgery protection to the AJAX section update actions in Pegasus.  (We will be migrating these actions to dashboard but having this protection in place prevents attacks in the meanwhile.)

We make the appropriate calls in the page header to automatically set the CSRF token on all Ajax requests and update the `rack-csrf` configuration to protect posts to the section routes.  Having this framework in place makes it easy to protect additional POST actions should we choose.

In addition to unit tests, I manually verified that the teacher dashboard ajax requests include the CSRF token and continue to work correctly.

Deployment note:  during the brief window of a push, some clients could submit requests without CSRF tokens that will be rejected.  This will succeed on retry once frontends are updgraded.  Because we deploy after the end of the school day, I think the number of affected users should be small so I think this is acceptable fallout.

### CODE REVIEW

**Comment by philbogle** [comment]
@wjordan FYI

File: pegasus/sites.v3/code.org/views/header.haml [hunk 1]
```diff
@@ -97,6 +97,13 @@
   var headerSetup = false;
 
   $ = jQuery;
+
+  $.ajaxSetup({
+    beforeSend: function(xhr) {
+      xhr.setRequestHeader('X_CSRF_TOKEN', '#{csrf_token}');
```
Comment (laurelfan) [inline pegasus/sites.v3/code.org/views/header.haml:103]: is this ever included in cached pages?

**Comment by laurelfan** [comment]
Is the sinatra session cookie a new cookie or is it an existing cookie that is already whitelisted for the appropriate paths in the caching config?

**Comment by philbogle** [comment]
Brilliant point.  It needs to be whitelisted, I will do that.  I didn't see
this in development because our cache simulation middleware seems to apply
only to dashboard not pegasus.

**Comment by wjordan** [comment]
> I didn't see this in development because our cache simulation middleware seems to apply only to dashboard not pegasus.

looking at the code, the cache simulation middleware is currently applied to pegasus as well as dashboard `config.ru` settings, but it's currently applied only to :development and not :test, which is perhaps why it didn't get invoked by the tests created.

File: pegasus/sites.v3/code.org/views/header.haml (hunk 1 above)
Comment (philbogle) [inline pegasus/sites.v3/code.org/views/header.haml:103]: The cached pages don't hit the CSRF protected teacher dashboard actions, and setting this elsewhere should be harmless.  However, I agree, I should look for a less global place to include this.  I will do that.

**Comment by philbogle** [comment]
Shoudn't we apply the cache simulation middleware in test also?

**Comment by wjordan** [comment]
maybe, though i'm not 100% sure- if there are tests that depend on the assumption that pages will not be cached then they could break, so it would need a full round of testing to make sure everything still works correctly.

also, keep in mind that we would need extra logic to distinguish between `test` meaning 'running tests' and `test` meaning 'pre-prod test server', since we definitely do not want to add the middleware to the latter.

an alternative approach would be to construct the Rack application manually in the unit test rather than loading through `config.ru` directly (or just manually adding the necessary middlewares to the Rack app after loading from `config.ru`).

**Comment by philbogle** [comment]
While caching is undesirable, it seems like cookie stripping is always
desirable (?)

**Comment by laurelfan** [comment]
This LGTM. It should be possible to manually test this in staging because staging has the caching/cookie setup like prod.

**Pull request merged**